libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,222 @@
+            from uuid import uuid4
+
+            from langchain_core.documents import Document
+            from langchain_core.indexes import AddResponse, DeleteResponse, Index
+
+            def uuid4_generator() -> Iterable[str]:
+                while True:
+                    yield str(uuid4())
+
+            class DictIndex(Index):
+
+                def __init__(self) -> None:
+                    self.store = {}
+
+                def add(
+                    self,
+                    documents: Iterable[Document],
+                    *,
+                    ids: Optional[Iterable[str]] = None,
+                    **kwargs: Any,
+                ) -> AddResponse:
+                    ids = ids or uuid4_generator()
+                    succeeded = []
+                    for id_, doc in zip(ids, documents):
+                        self.store[id_] = doc
+                        succeeded.append(id_)
+                    return AddResponse(succeeded=succeeded, failed=[])
+
+                def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+                    succeeded = []
+                    failed = []
+                    for id_ in ids:
+                        try:
+                            del self.store[id_]
+                        except Exception:
+                            failed.append(id_)
+                        else:
+                            succeeded.append(id_)
+                    return DeleteResponse(succeeded=succeeded, failed=failed)
+
+                def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+                    for id in ids:
+                        yield self.store[id]
+
+                def yield_keys(
+                    self, *, prefix: Optional[str] = None
+                ) -> Union[Iterator[str]]:
+                    prefix = prefix or ""
+                    for key in self.store:
+                        if key.startswith(prefix):
+                            yield key
+    """  # noqa: E501
+
+    @abstractmethod
+    def add(
+        self,
+        documents: Iterable[Document],
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> AddResponse:
+        """Add documents to index."""
+
+    @abstractmethod
+    def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+        """Delete documents by id.
+
+        Args:
+            ids: IDs of the documents to delete.
+
+        Returns:
+           A dict ``{"succeeded": [...], "failed": [...]}`` with the IDs of the
+           documents that were successfully deleted and the ones that failed to be
+           deleted.
+        """
+
+    @abstractmethod
+    def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+        """Lazily get documents by id.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Yields:
+           Document
+        """
+
+    def get_by_ids(self, ids: Iterable[str]) -> List[Document]:
```

Comment:
Sequence instead of Iterable perhaps to require a user to think about pagination